services/device-registry-base/src/main/java/org/eclipse/hono/deviceregistry/service/deviceconnection/MapBasedDeviceConnectionService.java

```diff
@@ -105,17 +111,151 @@ public MapBasedDeviceConnectionsConfigProperties getConfig() {
+            final JsonObject deviceAdapterInstanceIdJson = commandHandlersForTenantMap.get(deviceId);
+            if (deviceAdapterInstanceIdJson != null) {
+                // found mapping for given device id
+                resultMap.put(deviceId, getAdapterInstanceIdFromJson(deviceAdapterInstanceIdJson));
+            } else if (!viaGateways.isEmpty()) {
+                // no mapping found for given device; check last known gateway of device
+                final Map<String, JsonObject> lastKnownGatewaysForTenantMap = lastKnownGatewaysMap.get(tenantId);
+                if (lastKnownGatewaysForTenantMap != null) {
+                    final JsonObject lastKnownGatewayJson = lastKnownGatewaysForTenantMap.get(deviceId);
+                    if (lastKnownGatewayJson != null) {
+                        final String gatewayId = getGatewayIdFromLastKnownGatewayJson(lastKnownGatewayJson);
+                        if (viaGateways.contains(gatewayId)) {
+                            // get command handler for found gateway device
+                            final JsonObject gwAdapterInstanceIdJson = commandHandlersForTenantMap.get(gatewayId);
+                            if (gwAdapterInstanceIdJson != null) {
+                                resultMap.put(gatewayId, getAdapterInstanceIdFromJson(gwAdapterInstanceIdJson));
+                            }
+                        } else {
+                            log.trace("ignoring found last known gateway [{}]; gateway is not in given via list", gatewayId);
+                        }
+                    }
+                }
+            }
+            if (resultMap.isEmpty() && !viaGateways.isEmpty()) {
+                log.trace("no command handling adapter instance found for given device or last known gateway; getting instances for all via gateways");
+                for (final String viaGateway : viaGateways) {
+                    final JsonObject gwAdapterInstanceIdJson = commandHandlersForTenantMap.get(viaGateway);
+                    if (gwAdapterInstanceIdJson != null) {
+                        resultMap.put(viaGateway, getAdapterInstanceIdFromJson(gwAdapterInstanceIdJson));
+                    }
+                }
+            }
+            if (!resultMap.isEmpty()) {
+                result = DeviceConnectionResult.from(HttpURLConnection.HTTP_OK, getResultJson(resultMap));
+            } else {
+                result = DeviceConnectionResult.from(HttpURLConnection.HTTP_NOT_FOUND);
+            }
+        } else {
+            result = DeviceConnectionResult.from(HttpURLConnection.HTTP_NOT_FOUND);
+        }
+        return Future.succeededFuture(result);
+    }
+
+    private JsonObject getResultJson(final Map<String, String> deviceToAdapterInstanceMap) {
+        final JsonObject jsonObject = new JsonObject();
+        final JsonArray adapterInstancesArray = new JsonArray(new ArrayList<>(deviceToAdapterInstanceMap.size()));
```

Comment:
IMHO this should simply be
```
final JsonArray adapterInstancesArray = new JsonArray();
```

Reply:
Since the array size is fixed here, I want to initialize the JsonArray with that size like that (for lack of a corresponding JsonArray constructor).